src/search.cpp

```diff
@@ -1457,6 +1498,92 @@ namespace {
     return best;
   }
 
+
+  // probe_tb() is called by search after TT probe to look up the position in
+  // tablebases, if it is found we return immediately or continue the search
+  // eventually with reduced depth according to the case.
+
+  Value probe_tb(Position& pos, Stack* ss, Value alpha, Value beta, Depth* depth, TTEntry* tte) {
+
+    TB::ProbeState err;
+    TB::WDLScore wdl;
+    bool tbHit = false, farFromRoot = false, reduceDepth = true;
+
+    Thread* thisThread = pos.this_thread();
+    thisThread->tbHits.fetch_add(1, std::memory_order_relaxed);
+
+    // If rootPos is in TB we probe DTZ to differentiate between a win/loss and
+    // a cursed/blessed draw. This is critical because WDL is reliable only if
+    // rule50 counter is zero and search alone would detect it only at ply 101.
+    if (TB::CanProbeDTZ && (ss-1)->ply == 1)
+    {
+        tbHit = true;
+        RootMove& rm = *std::find(thisThread->rootMoves.begin(), thisThread->rootMoves.end(),
+                                  (ss-1)->currentMove);
+        wdl = TB::dtz_to_wdl(rm.dtz, pos.rule50_count());
```

Comment:
On this line the second argument should be the 50-move-counter of the root position, not the 50-move-counter of the current position. (So the field r50 field would have been useful after all, sorry!)

The way that you are now computing rm.dtz is clearly a lot nicer but note that it is a rewrite of code in the current master:

https://github.com/official-stockfish/Stockfish/blob/master/src/syzygy/tbprobe.cpp#L1540-L1573

There are a couple of things that make me worry now:

1. The special test for checkmates in current master (tbprobe.cpp lines 1546-1551) is now gone. This could be just an academic issue because I doubt Stockfish is going to miss mates in 1. But if you are passing data from tbprobe.cpp to search.cpp it would be nice if all the data was correct.

2. Regarding my idea that you can test for optimality by checking RootPosDTZ == rm.dtz. I think in an ideal world that would be correct however the probe_dtz function comes with a warning:

```
// The return value n can be off by 1: a return value -n can mean a loss
// in n+1 ply and a return value +n can mean a win in n+1 ply. This
// cannot happen for tables with positions exactly on the "edge" of
// the 50-move rule.
```

Maybe this explains why @syzygy1 computes the best DTZ of all root moves:

https://github.com/official-stockfish/Stockfish/blob/master/src/syzygy/tbprobe.cpp#L1603-L1610

In an ideal world I think the variable 'best' computed here would always be equal to the variable 'dtz' computed on line 1533 but the way the probe_dtz function works maybe these two values can be off by one? I don't know, this is an untested theory. I'm just slightly nervous that there can be cases where no root moves satisfy the condition RootPosDTZ == rm.dtz.

Reply:
Should we set and use a TB::Root50 variable?

Reply:
@ceebo I have updated the PR, please check it.